turn/src/server/mod.rs

```diff
@@ -52,43 +56,63 @@ impl Server {
             let auth_handler = Arc::clone(&s.auth_handler);
             let realm = s.realm.clone();
             let channel_bind_timeout = s.channel_bind_timeout;
-            let shutdown_rx = shutdown_rx.clone();
-
-            tokio::spawn(async move {
-                let allocation_manager = Arc::new(Manager::new(ManagerConfig {
-                    relay_addr_generator: p.relay_addr_generator,
-                }));
-
-                Server::read_loop(
-                    p.conn,
-                    allocation_manager,
-                    nonces,
-                    auth_handler,
-                    realm,
-                    channel_bind_timeout,
-                    shutdown_rx,
-                )
-                .await;
+            let handle_rx = handle.subscribe();
+            let conn = p.conn;
+            let allocation_manager = Arc::new(Manager::new(ManagerConfig {
+                relay_addr_generator: p.relay_addr_generator,
+            }));
+
+            tokio::spawn({
+                let allocation_manager = Arc::clone(&allocation_manager);
+
+                async move {
+                    Server::read_loop(
+                        conn,
+                        allocation_manager,
+                        nonces,
+                        auth_handler,
+                        realm,
+                        channel_bind_timeout,
+                        handle_rx,
+                    )
+                    .await;
+                }
             });
         }
 
         Ok(s)
     }
 
+    /// Deletes all existing [`crate::allocation::Allocation`]s by the provided `username`.
+    pub async fn delete_allocations_by_username(&self, username: String) -> Result<()> {
+        let tx = self.handle.lock().await.clone();
+        if let Some(tx) = tx {
+            let (closed_tx, closed_rx) = mpsc::channel(1);
+            tx.send(Command::DeleteAllocations(username, Arc::new(closed_rx)))
+                .map_err(|_| Error::ErrClosed)?;
+
+            closed_tx.closed().await;
+
+            Ok(())
+        } else {
+            Err(Error::ErrClosed)
+        }
+    }
+
     async fn read_loop(
         conn: Arc<dyn Conn + Send + Sync>,
         allocation_manager: Arc<Manager>,
         nonces: Arc<Mutex<HashMap<String, Instant>>>,
         auth_handler: Arc<dyn AuthHandler + Send + Sync>,
         realm: String,
         channel_bind_timeout: Duration,
-        mut shutdown_rx: watch::Receiver<bool>,
+        mut handle_rx: broadcast::Receiver<Command>,
     ) {
         let mut buf = vec![0u8; INBOUND_MTU];
 
         loop {
-            let (n, addr) = tokio::select! {
-                v = conn.recv_from(&mut buf) => {
+            let (n, addr) = futures::select! {
```

Comment:
@k0nserv as `tokio`'s select has [unsafe cancellation](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety), we decided to use `future`'s to avoid losing messages.

Reply:
`future`'s implementation suffer from the same problem, see https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=74a26199b352aa8b3403916c6fb990f9

Any select over futures with more than a single await point have to consider cancel safety, because it's always possible that some progress was made in the branch not taken and then dropped

Reply:
@k0nserv it seems you are right. So, I'll think how to solve it.

Reply:
@k0nserv I have reworked this.